src/ore/src/channel.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Channel utilities and extensions.
+
+use std::collections::VecDeque;
+
+use async_trait::async_trait;
+
+/// Extensions for the receiving end of asynchronous channels.
+#[async_trait(?Send)]
+pub trait ReceiverExt<T> {
+    /// Receives all of the currently buffered elements on the channel, up to some max.
+    ///
+    /// This method returns `None` if the channel has been closed and there are no remaining
+    /// messages in the channel's buffer.
+    ///
+    /// If there are no messages in the channel's buffer, but the channel is not yet closed, this
+    /// method will sleep until a message is sent or the channel is closed. When woken it will
+    /// return up to max currently buffered elements.
+    async fn recv_many(&mut self, max: usize) -> Option<VecDeque<T>>;
+}
+
+#[async_trait(?Send)]
+impl<T> ReceiverExt<T> for tokio::sync::mpsc::Receiver<T> {
+    async fn recv_many(&mut self, max: usize) -> Option<VecDeque<T>> {
+        // Wait for a value to be ready.
+        let first = self.recv().await?;
+        let mut buffer = VecDeque::from([first]);
```

Comment:
why does this use a `VecDeque` if you only ever call `push_back` on it? seems like it'd be simpler as just a `Vec`

Reply:
Theoretically a caller might want to consume the values in the same order in which they were received, for which they can call `pop_front(...)`

Reply:
converting `Vec` into `VecDeque` is a guaranteed cheap operation though, and i would think people typically would just want to iterate over it anyway

Reply:
TIL `Vec` into `VecDeque` is `O(1)` that's great! I'll update this do use `Vec`